### ISSUE TITLE
Runtime error - unresolved external symbols for Zlib Compress/DeCompress

### ISSUE THREAD

**Issue by pmahend1** [kareldonk/QuantumGate#1]
Getting errors like below 

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol _inflateInit_@12 referenced in function "public: static bool const __cdecl QuantumGate::Implementation::Compression::ZlibStreams::Decompress(enum std::byte *,unsigned int &,class QuantumGate::Implementation::Memory::BufferView const &)" (?Decompress@ZlibStreams@Compression@Implementation@QuantumGate@@SA?B_NPAW4byte@std@@AAIABVBufferView@Memory@34@@Z)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _inflateEnd@4 referenced in function "public: void __thiscall <lambda_15208b9786da0bdb63571e938c20c530>::operator()(void)const " (??R<lambda_15208b9786da0bdb63571e938c20c530>@@QBEXXZ)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _inflate@8 referenced in function "public: static bool const __cdecl QuantumGate::Implementation::Compression::ZlibStreams::Decompress(enum std::byte *,unsigned int &,class QuantumGate::Implementation::Memory::BufferView const &)" (?Decompress@ZlibStreams@Compression@Implementation@QuantumGate@@SA?B_NPAW4byte@std@@AAIABVBufferView@Memory@34@@Z)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _deflateInit_@16 referenced in function "public: static bool const __cdecl QuantumGate::Implementation::Compression::ZlibStreams::Compress(enum std::byte *,unsigned int &,class QuantumGate::Implementation::Memory::BufferView const &)" (?Compress@ZlibStreams@Compression@Implementation@QuantumGate@@SA?B_NPAW4byte@std@@AAIABVBufferView@Memory@34@@Z)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _deflateEnd@4 referenced in function "public: void __thiscall <lambda_d0a06043170a546814590c54b4a9b3c6>::operator()(void)const " (??R<lambda_d0a06043170a546814590c54b4a9b3c6>@@QBEXXZ)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _deflate@8 referenced in function "public: static bool const __cdecl QuantumGate::Implementation::Compression::ZlibStreams::Compress(enum std::byte *,unsigned int &,class QuantumGate::Implementation::Memory::BufferView const &)" (?Compress@ZlibStreams@Compression@Implementation@QuantumGate@@SA?B_NPAW4byte@std@@AAIABVBufferView@Memory@34@@Z)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK2019	unresolved external symbol _compressBound@4 referenced in function "bool const __cdecl QuantumGate::Implementation::Compression::Compress(class QuantumGate::Implementation::Memory::BufferView const &,class QuantumGate::Implementation::Memory::BufferImpl<class QuantumGate::Implementation::Memory::PoolAllocator<enum std::byte> > &,enum QuantumGate::Algorithm::Compression)" (?Compress@Compression@Implementation@QuantumGate@@YA?B_NABVBufferView@Memory@23@AAV?$BufferImpl@V?$PoolAllocator@W4byte@std@@@Memory@Implementation@QuantumGate@@@523@W41Algorithm@3@@Z)	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Examples\QuantumGateConsole\QuantumGate32D.lib(Compression.obj)	1	
Error	LNK1120	7 unresolved externals	QuantumGateConsole	F:\Project\ProjectNet\QuantumGate\Build\Win32\Debug\QuantumGateConsoleExample.exe	1	

```

This error is coming when trying to run `CmdMess` or `QuantumGateConsole` error,

similar error was throwing while building `QuantumGateLib` project by changing output type to `lib` from `different options` and adding `zlibwapi.dll` and `zlibwapi.lib` files solved it , but the lib produced in turn throwing same error.

Seems it has to do with `zlibwapi` i.e `ZLIB_WINAPI` macro as I read in some posts.

Please help fixing this issue.

**Comment by kareldonk** [comment]
Could you provide some more information? What did you use to build the zlib DLL? What did you use, and how did you build the QuantumGateLib project? What OS/ platform?

Note from the [ReadMe](https://github.com/kareldonk/QuantumGate#build) on the main GitHub page that the QuantumGateLib project is configured to look for specific names for the zlib library files (.DLL and .lib). You need to build the zlib project to output those file names, or else change the configuration. The filenames for the external library dependencies can be found in the [stdafx.h](https://github.com/kareldonk/QuantumGate/blob/master/QuantumGateLib/stdafx.h) file.

**Comment by kareldonk** [comment]
I don't think you can rename the zlib `.dll` and `.lib` files *after* building them. You need to specify the correct output filenames *before* building them.

The way I build the zlib library, is by using the Visual Studio solution `zlibvc.sln` from the `contrib\vstudio\vc14` folder in the root zlib project folder. If you open that in Visual Studio 2017, you then need to build the `zlibvc` project in 32-bit and 64-bit **Release** configuration. But **before** you issue the build command in Visual Studio for that project, you need to go into the settings by right-clicking on the `zlibvc` project in Solution Explorer, then choosing Properties from the menu. Then go to Linker - All Options and then change all the output filenames from `zlibwapi.*` to `zlib32.*` for 32 bit platform and `zlib64.*` for 64 bit platform. Then build. Use the output `.dll` and `.lib` files to build the QuantumGate projects.

**Comment by pmahend1** [comment]
Will try that and update you.

**Comment by pmahend1** [comment]
ok.I commented out line 44 i.e 

`assert(BCryptAlgorithm != NULL)` , now it wont error. Hope that was expected.

**Comment by pmahend1** [comment]
QuantumGate was `Lib` file. Used same lib/dll files for other dependencies of Example projects

**Comment by pmahend1** [comment]
In debug mode assert statement crashes. Release mode I uncommented it, its not crashing. Am I supposed to run all projects in Release mode?

**Comment by kareldonk** [comment]
I think you did not correctly build the QuantumGate library. 

You should build the QuantumGate library as a dynamic library (`.dll` with accompanying `.lib`). Building it as a static library (only `.lib`) currently is not (yet) supported. 

If you open the `QuantumGate.sln` file in the root folder in Visual Studio 2017, you can simply select your configuration (Debug or Release) and platform (Win32 or x64) from the toolbar and then issue the build command in Visual Studio. There's no need to change any settings, except for making sure that the dependency files (json, monocypher, zlib etc.) are in the Visual Studio include paths otherwise you will get build errors where files cannot be found.

**Comment by pmahend1** [comment]
That solved it.Thanks 👍

**Comment by kareldonk** [comment]
Great, thanks. Remember not to remove assertions in debug mode (or release); they are there for a good reason. I'll be closing this issue if you don't have any further questions. You can always open a new one.

**Comment by pmahend1** [comment]
Running in release mode , I am able to run the TestApp exe but cant connect two instances on same machine or another machine even after turning off firewall. It says accepting connections but then Disconnecting soon after, `Operation failed. Not allowed by security configuration.";` this is not related to library issue.Seems config issue. Any suggestions?

**Comment by kareldonk** [comment]
That is related to the security settings for QuantumGate, as configured by the TestApp. For all the details about QuantumGate security settings, you can look at [`AccessManager`](https://github.com/kareldonk/QuantumGate/wiki/QuantumGate::AccessManager). The TestApp also uses the same API to configure QuantumGate. By default QuantumGate is locked down as much as possible to prevent unintended use and access.

Specific to TestApp, if you go in the `Local` menu on top of the app, you will find `IPFilters`, `IPSubnet Limits` and `Peer Access Settings` menu options.
You need to (1) white list the IP addresses of your computers in IPFilters (you can use a mask or range), then you need to (2) add the peers that are allowed to connect via menu option Peer Access Settings - Add/Remove Peers. You can skip (2) if you disable menu option Peer Access Settings - Require Authentication. You also (3) need to go into Peer Access Settings - Add/Remove Peers, and in the Peer Access Settings dialog, you need to put the option "default access for peers not in this list" to Allowed. Keep in mind that this would allow all peers to connect. But for testing and tryout this is not an issue.

**Comment by pmahend1** [comment]
Yes. we figured that out and added pem file to accept connection. Even file transfer works . Thanks 👍

**Comment by kareldonk** [comment]
Excellent
